src/Agent.Plugins/PipelineCache/PipelineCacheServer.cs

```diff
@@ -151,18 +178,41 @@ public class PipelineCacheServer
             return pipelineCacheClient;
         }
 
-        private Task DownloadPipelineCacheAsync(
+        private async Task<string> GetUploadPathAsync(string preferredContentFormat, AgentTaskPluginExecutionContext context, string path, CancellationToken cancellationToken)
+        {
+            string uploadPath = path;
+            if(string.Equals(preferredContentFormat, ContentFormatConstants.SingleTar, StringComparison.OrdinalIgnoreCase))
+            {
+                uploadPath = await TarUtils.ArchiveFilesToTarAsync(context, path, cancellationToken);
+            }
+            return uploadPath;
+        }
+
+        private async Task DownloadPipelineCacheAsync(
+            AgentTaskPluginExecutionContext context,
             DedupManifestArtifactClient dedupManifestClient,
             DedupIdentifier manifestId,
             string targetDirectory,
+            string contentFormat,
             CancellationToken cancellationToken)
         {
-            DownloadDedupManifestArtifactOptions options = DownloadDedupManifestArtifactOptions.CreateWithManifestId(
-                manifestId,
-                targetDirectory,
-                proxyUri: null,
-                minimatchPatterns: null);
-            return dedupManifestClient.DownloadAsync(options, cancellationToken);
+            if (contentFormat == ContentFormatConstants.SingleTar)
+            {
+                string manifestPath = Path.Combine(Path.GetTempPath(), $"{nameof(DedupManifestArtifactClient)}.{Path.GetRandomFileName()}.manifest");
```

Comment:
>manifestPath [](start = 23, length = 12)

cleanup

Reply:
better yet, just use DownloadToStreamAsync to download it to a memorystream

---
In reply to: [315357934](https://github.com/microsoft/azure-pipelines-agent/pull/2369#discussion_r315357934) [](ancestors = 315357934)

Reply:
That was the plan, but @owenhuynMSFT had some doubts regarding streaming and JSON Deserializer not being happy. https://github.com/microsoft/azure-pipelines-agent/pull/2369#discussion_r313133183